Adding RAFT IVF-PQ and IVF-Flat

cjnolet opened the pull request.

cjnolet commented:
@erikbern is there interest in having this algorithm included in ann-benchmarks? We have a couple new GPU-accelerated ANN implementations as part of the Nvidia RAPIDS RAFT ecosystem. We're working to integrate these into FAISS but are also finding value in being able to benchmark them individually for comparison. If there's interest in adding the algorithms from the RAPIDS RAFT library, I can finish doing the integration in this PR.

erikbern commented on README.md:
```diff
@@ -7,6 +7,39 @@ Doing fast searching of nearest neighbors in high dimensional spaces is an incre
 
 This project contains some tools to benchmark various implementations of approximate nearest neighbor (ANN) search for different metrics. We have pregenerated datasets (in HDF5) formats and we also have Docker containers for each algorithm. There's a [test suite](https://travis-ci.org/erikbern/ann-benchmarks) that makes sure every algorithm works.
 
+
```
Can you please remove this, since (a) I don't think it belongs in the README (b) we use Docker for everything, not Conda

cjnolet replied:
Absolutely I'll remove this. It's only temporary to help others (and myself) during evaluation.

erikbern commented:
I don't run the benchmarks on GPUs right now and I'm not sure if we'll do that in the future – since the cost is so much higher. What's the speed up on GPUs over CPUs?

erikbern commented:
Would love to merge this if it's competitive though! I want ann-benchmarks to have as many algorithms as possible!

erikbern commented on ann_benchmarks/main.py:
```diff
@@ -122,6 +123,12 @@ def main():
+    parser.add_argument(
+        '--batch-size',
+        help='Number of vectors in each batch. only works in batch mode. default (-1) is a single batch',
+        required=True,
```
Can you make this not required? All the tests are failing because of this right now. Thanks!

cjnolet commented:
> I don't run the benchmarks on GPUs right now and I'm not sure if we'll do that in the future – since the cost is so much higher. What's the speed up on GPUs over CPUs?

That's fair. There might be something we can work out on the RAPIDS side where we clone this repository and run a few of the GPU-based tests nightly or something along those lines. This is of high interest to us on the RAPIDS RAFT side currently. 

We are seeing some good speedup using our implementations, even over SCaNN for small(-ish) batch sizes. I added the `batch-size` option to this repository so that we can evaluate the space in between `batch_size=len(X_test)` and `batch_size=1`. So far at least the preliminary results are looking promising, especially as RAFT's 4bit PQ + 8bit LUT + refinement appears to be consistently pushing the QPS over SCaNN at higher recall. I'll share some preliminary slides (created w/ the changes in this PR) soon. 

We are also working on a [graph-based ANN on the GPU](https://github.com/rapidsai/raft/issues/997) which is showing even better performance for small batch sizes.

cjnolet commented:
Moving instructions to PR so they can removed from the code changes:

# Running RAFT Benchmarks From Start to Finish on SIFT-1M

To benchmark SCaNN:
```bash
# To benchmark scann
mamba create --name ann_benchmarks_scann python=3.6
conda activate ann_benchmarks_scann
pip install -r requirements.txt
pip install scann
python run.py --batch --local --dataset=sift-128-euclidean --algorithm scann --batch-size 1000
conda deactivate
```

To benchmark RAFT:
```bash
# To benchmark RAFT
mamba create --name ann_benchmarks_raft python=3.10
mamba install -c conda-forge -c rapidsai -c nvidia pylibraft==23.02 rmm scipy h5py psutil pyyaml cupy
pip install docker ansicolors
python run.py --batch --local --dataset=sift-128-euclidean --algorithm=raft-ivfpq --run-disabled --batch-size 1000
conda deactivate
```

To Plot results:
```bash
conda activate ann_benchmarks_scann
python plot.py --batch --dataset sift-128-euclidean -o sift-128-euclidean.png
open sift-128-euclidean.png
```

cjnolet commented:
I've got the rapids docker container working for me locally but it did require that I add (optional)  `Definition` properties for the docker runtime and additional environment options. For now, I've set the raft algorithm to disabled but I'm going to look into what it will take to get the container running nightly in the RAPIDS CI.

maumueller commented on ann_benchmarks/results.py:
```diff
@@ -44,6 +44,7 @@ def store_results(dataset, count, definition, query_arguments, attrs, results,
 
 def load_all_results(dataset=None, count=None, batch_mode=False):
     for root, _, files in os.walk(get_result_filename(dataset, count)):
+        print(str(files))
```
Please remove.

maumueller commented on ann_benchmarks/runner.py:
```diff
@@ -137,13 +148,18 @@ def run(definition, dataset, count, run_count, batch):
             descriptor["algo"] = definition.algorithm
             descriptor["dataset"] = dataset
+
+            print(str(descriptor))
```
Remove.

maumueller commented on ann_benchmarks/runner.py:
```diff
@@ -137,13 +148,18 @@ def run(definition, dataset, count, run_count, batch):
                           query_arguments, descriptor, results, batch)
+    except:
+        print("FAILED!")
+        print(str(e))
```
Since this seems to have been important in your testing, maybe make this prettier but keep it?

maumueller commented:
Hi @cjnolet! `batchsize` is a nice feature that I would also like to see included. Your solution of using a different dockerfile as basis to run gpu experiments seems fine as well. There are many debug print statements, please remove those for the final version. 

Some general thoughts: I added the `batch` support many years ago to allow targeting GPU/"exotic" architectures. FAISS' ivf worked extremely well for me and I used it back in 2018 when writing up [our paper](https://arxiv.org/pdf/1807.05614.pdf).

![image](https://user-images.githubusercontent.com/6311646/221852899-142a1f08-2454-4a2e-99c6-b269261d5d39.png)

I suppose the difference in throughput only increased by now. It's a bit apples and oranges though, so it's difficult to present results in a single meaningful plot as I tried above. I would be happy with a more open leader board, but it should be clearly distinguished from our  single-threaded CPU-based runs. 

Side note: The "sister project" targeting out-of-memory billion-scale ANN had a track for non-traditional settings with a leaderboard: <https://github.com/harsha-simhadri/big-ann-benchmarks/blob/main/t3/LEADERBOARDS_PUBLIC.md> It was the most interesting track in our NeurIPS challenge with some amazing results. :-) @sourcesync did all the hard work of evaluating the effort for this.

cjnolet commented:
> Side note: The "sister project" targeting out-of-memory billion-scale ANN had a track for non-traditional settings with a leaderboard: https://github.com/harsha-simhadri/big-ann-benchmarks/blob/main/t3/LEADERBOARDS_PUBLIC.md It was the most interesting track in our NeurIPS challenge with some amazing results. :-) 

@maumueller funny enough, RAFT's IVFPQ implementation is the `cuanns_ifvpq` implementation on that leaderboard. I've removed the unneeded prints. Indeed RAFT's new ANN implementations are optimized for the smaller batch size and we're trying hard to be competitive in that realm. We also have a new graph-based algorithm coming out soon that we're finding is very competitive with SCaNN and HNSW for smaller batch sizes.

Question: What batch sizes do you often see in production. I'm hearing they can range from 5-3000 but it seems the smaller sizes are more common. We're also trying to make sure we're benchmarking the most realistic cases.

cjnolet commented on ann_benchmarks/runner.py:
```diff
@@ -263,12 +288,15 @@ def stream_logs():
-    if type(return_value) is dict: # The return value from container.wait changes from int to dict in docker 3.0.0
-        error_msg = return_value['Error']
+    # The return value from container.wait changes from int to dict in docker 3.0.0
+    if type(return_value) is dict:
```
I'm not sure if something has changed recently in the `docker-py` library but it looks like a dict is always returned now and it doesn't always contain an `Error` key.

erikbern replied:
yeah the docker library has changed a few times with similar things, it's a bit annoying

cjnolet commented on install/Dockerfile.raft:
```diff
@@ -0,0 +1,10 @@
+
+COPY --from=ann-benchmarks /home/app /home/app
+
+ENTRYPOINT ["python3", "-u", "/home/app/run_algorithm.py"]
```
It looks like the final results are still in the container after running the benchmarks. I'm trying to figure out a good way to copy them back onto the local filesystem.

cjnolet replied:
I *think* I figured this out. It looks like it works when I set the `WORKDIR` here, though I'm not a docker expert

maumueller replied:
We mount-bind  the host's `results/` as read-write to `/home/app/results`.

cjnolet commented:
@erikbern @maumueller It looks like RAFT's 4-bit IVF-PQ implementation is pretty competitive to scann so far. I ran this benchmark on an A100 GPU w/ an Intel Xeon 24-core (48 threads) for batch-size 10:

![image](https://user-images.githubusercontent.com/1242464/223311255-d7ee30b8-dacf-4758-a5c7-5282c02ab41d.png)

I think this PR is just about ready. Please let me know if there's anything else I can do (or fix).

erikbern commented on algos.yaml:
```diff
@@ -58,6 +58,18 @@ float:
           args: [[400, 1024, 4096, 8192, 16384],
                  [1, 10, 40, 100, 200]]
+    raft-ivfpq:
+      disabled: true
```
why is it disabled?

cjnolet replied:
Only because it requires a GPU and I wasn't sure if it's okay to assume one will always be available.

erikbern replied:
ok that makes sense! yeah let's keep it disabled then

erikbern commented:
Do you want to add it to https://github.com/erikbern/ann-benchmarks/blob/master/.github/workflows/benchmarks.yml so that we can include it in the CI runs?

erikbern commented:
Also curious how it fares with batchsize = 1? I'm about to re-run all benchmarks, so it would be great to include this library

cjnolet commented:
> Do you want to add it to https://github.com/erikbern/ann-benchmarks/blob/master/.github/workflows/benchmarks.yml so that we can include it in the CI runs?

Sure, but RAFT does require a GPU. Does CI have GPU resources available?

cjnolet commented:
> Also curious how it fares with batchsize = 1? I'm about to re-run all benchmarks, so it would be great to include this library

It does fair well with batch-size = 1 but I'm not sure it's the most fair comparison. I had enabled the batch-size option mostly to allow the CPU to run w/ more threads. It seems like once we hit the CPU multi-threading limit on the machine, the GPU dominates.

erikbern commented:
We haven't run any benchmarks on GPUs so far but I'm open to it! Curious what speedups you see?

cjnolet commented:
We did recently enable GPU support in Github Actions for the Nvidia RAPIDS libraries (the ecosystem for which the RAFT library is a part of). I think for now we're talking about running the GPU ann-benchmarks nightly on the RAPIDS side and maybe we can see about eventually GPU-enabling this repository as well.

erikbern commented:
is this ready to be merged?

cjnolet commented:
@erikbern I think so. It looks like the CI is failing for reasons that don't seem related to the changes in this PR. Is that expected?

sourcesync commented:
Hey @cjnolet  i'm getting these errors when I try to run just this algorithm:

'''
$ python run.py --dataset deep-image-96-angular --algorithm raft-ivfpq 
...
2023-03-13 00:57:23,612 - annb - INFO - missing docker images: {'ann-benchmarks-raft'}
Traceback (most recent call last):
  File "run.py", line 6, in <module>
    main()
  File "/home/gwilliams/Projects/ann-benchmarks/cjnolet/ann-benchmarks/ann_benchmarks/main.py", line 236, in main
    raise Exception('Nothing to run')
Exception: Nothing to run
'''

If I try to build the containers, I get the following error:

'''
$ python install.py --algorithm raft-ivfpq
...
...
Unpacking manpages-dev (5.10-1ubuntu1) ...
Killed
The command '/bin/sh -c apt-get install -y python3-numpy python3-scipy python3-pip build-essential git' returned a non-zero code: 137
Traceback (most recent call last):
  File "install.py", line 49, in <module>
    subprocess.check_call(
  File "/home/gwilliams/anaconda3/envs/ann-benchmarks-cjnolet/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'docker build         --rm -t ann-benchmarks -f install/Dockerfile .' returned non-zero exit status 137.

'''

Apologies in advance if I missed some key instruction.  Here are my system details:
'''
$ uname -a
Linux sv7-apu12 5.4.0-139-generic #156-Ubuntu SMP Fri Jan 20 17:27:18 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/issue
Ubuntu 20.04.5 LTS \n \l
$ docker version
Client: Docker Engine - Community
 Version:           20.10.16
 API version:       1.41
 Go version:        go1.17.10
 Git commit:        aa7e414
 Built:             Thu May 12 09:17:28 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.16
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.10
  Git commit:       f756502
  Built:            Thu May 12 09:15:33 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.4
  GitCommit:        212e8b6fa2f44b9c21b2798135fc6fb7c53efc16
 runc:
  Version:          1.1.1
  GitCommit:        v1.1.1-0-g52de29d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
$ nvidia-smi
Mon Mar 13 01:01:38 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.85.12    Driver Version: 525.85.12    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:3C:00.0 Off |                  N/A |
|  0%   26C    P8    14W / 220W |      6MiB /  8192MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      2330      G   /usr/lib/xorg/Xorg                  4MiB |
+-----------------------------------------------------------------------------+

'''

cjnolet commented:
@sourcesync, 

That's strange, I'm getting an error just trying to run this:

```
(ann_benchmarks_py36) cjnolet@devgpu ~/workspace/rapids_projects/ann-benchmarks (add_raft_ivf_pq) $ python install.py --algorithm ann-benchmarks
Building base image...
[+] Building 0.8s (7/12)                                                                                                                                                                                           
 => [internal] load build definition from Dockerfile                                                                                                                                                          0.0s
 => => transferring dockerfile: 377B                                                                                                                                                                          0.0s
 => [internal] load .dockerignore                                                                                                                                                                             0.0s
 => => transferring context: 53B                                                                                                                                                                              0.0s
 => [internal] load metadata for docker.io/library/ubuntu:22.04                                                                                                                                               0.5s
 => [1/8] FROM docker.io/library/ubuntu:22.04@sha256:2adf22367284330af9f832ffefb717c78239f6251d9d0f58de50b86229ed1427                                                                                         0.0s
 => [internal] load build context                                                                                                                                                                             0.0s
 => => transferring context: 73B                                                                                                                                                                              0.0s
 => CACHED [2/8] RUN apt-get update                                                                                                                                                                           0.0s
 => ERROR [3/8] RUN apt-get install -y python3-numpy python3-scipy python3-pip build-essential git                                                                                                            0.3s
------
 > [3/8] RUN apt-get install -y python3-numpy python3-scipy python3-pip build-essential git:
#0 0.307 Reading package lists...
#0 0.314 Building dependency tree...
#0 0.315 Reading state information...
#0 0.315 E: Unable to locate package python3-numpy
#0 0.315 E: Unable to locate package python3-scipy
#0 0.315 E: Unable to locate package python3-pip
#0 0.315 E: Unable to locate package build-essential
#0 0.315 E: Unable to locate package git
------
Dockerfile:4
--------------------
   2 |     
   3 |     RUN apt-get update
   4 | >>> RUN apt-get install -y python3-numpy python3-scipy python3-pip build-essential git
   5 |     RUN pip3 install -U pip
   6 |     RUN python3 --version | grep 'Python 3.10.6'
--------------------
ERROR: failed to solve: process "/bin/sh -c apt-get install -y python3-numpy python3-scipy python3-pip build-essential git" did not complete successfully: exit code: 100
```

Are able install the base `Dockerfile` for ann-benchmarks?

sourcesync commented:
@cjnolet I'll have to investigate these memory issues this week, but it's related definitely related to docker.  I suggest holding off on this merge until we sort this out.

cjnolet commented:
@sourcesync sounds good. Let me know if there's anything I can do to help.

sourcesync commented:
Thanks @cjnolet.  I researched the issue and I've implemented a couple of work-arounds which seems to have gotten me past the main docker build issue.  Currently, I'm building all the docker containers...so far so good (too early to say if those workarounds were definitive or voodoo.)  When that is done, I'll run the benchmarks on just your algorithm.  This might take a while but I'll update asap when I have results good or bad.  ( BTW, I retrieved your latest files...)

cjnolet commented:
@sourcesync sounds good. I'm also attaching a PDF of the results I got comparing scann with `--batch` and various different settings for `--batch-size`. Unfortunately, RAFT's IVF-PQ comparison against SCANN for batch-size=1 isn't great. We do have a [graph-based GPU ANN](https://github.com/rapidsai/raft/issues/997) algorithm that's going to be integrated into RAFT very soon, which is showing promising performance for that case, though. 


[SCaNN vs RAFT IVF-PQ (recall@10) (2).pdf](https://github.com/erikbern/ann-benchmarks/files/10962491/SCaNN.vs.RAFT.IVF-PQ.recall%4010.2.pdf)

sourcesync commented:
@cjnolet So far so good.  It's running but may take a while to finish.  I'll put the graphs here when done.  BTW, i'm running this command:  
```python3 run.py --dataset deep-image-96-angular --algorithm raft-ivfpq --run-disabled```

cjnolet commented:
@sourcesync I want to let you know, we're seeing pretty poor performance for the single query at a time. I think a lot of it is related to the Python overheads because we're getting 11k+ QPS for single batch performance in C++ w/ the same parameters. This was also one of the reasons for adding the `--batch-size` option.

sourcesync commented:
@cjnolet After running all the query groups for your algorithm, it looks like I'm getting the following error:

```
python3 run.py --dataset deep-image-96-angular --algorithm raft-ivfpq --run-disabled
...
...
2023-03-15 17:24:10,997 - annb.0f15e234b9 - INFO - Run 5/5...
2023-03-15 17:24:11,998 - annb.0f15e234b9 - INFO - Processed 1000/10000 queries...
2023-03-15 17:24:13,001 - annb.0f15e234b9 - INFO - Processed 2000/10000 queries...
2023-03-15 17:24:14,004 - annb.0f15e234b9 - INFO - Processed 3000/10000 queries...
2023-03-15 17:24:15,005 - annb.0f15e234b9 - INFO - Processed 4000/10000 queries...
2023-03-15 17:24:16,010 - annb.0f15e234b9 - INFO - Processed 5000/10000 queries...
2023-03-15 17:24:17,012 - annb.0f15e234b9 - INFO - Processed 6000/10000 queries...
2023-03-15 17:24:18,015 - annb.0f15e234b9 - INFO - Processed 7000/10000 queries...
2023-03-15 17:24:19,017 - annb.0f15e234b9 - INFO - Processed 8000/10000 queries...
2023-03-15 17:24:20,019 - annb.0f15e234b9 - INFO - Processed 9000/10000 queries...
2023-03-15 17:24:21,019 - annb.0f15e234b9 - INFO - Processed 10000/10000 queries...
2023-03-15 17:24:24,331 - annb.0f15e234b9 - ERROR - Container.wait for container 0f15e234b9 failed with exception
Traceback (most recent call last):
  File "/home/gwilliams/Projects/ann-benchmarks/cjnolet/ann-benchmarks/ann_benchmarks/runner.py", line 288, in run_docker
    _handle_container_return_value(return_value, container, logger)
  File "/home/gwilliams/Projects/ann-benchmarks/cjnolet/ann-benchmarks/ann_benchmarks/runner.py", line 303, in _handle_container_return_value
    error_msg = 'Error' in return_value['Error']
TypeError: argument of type 'NoneType' is not iterable
2023-03-15 17:24:24,988 - annb.0a4e95f351 - INFO - Created container 0a4e95f351: CPU limit 1, mem limit 757792062208, timeout 7200, command ['--dataset', 'deep-image-96-angular', '--algorithm', 'raft-ivfpq', '--module', 'ann_benchmarks.algorithms.raft_ivfpq', '--constructor', 'RAFTIVFPQ', '--runs', '5', '--count', '10', '--batch-size', '0', '["angular", 400, 5, 64]', '[5, 0, "uint8", "float32"]', '[5, 0, "uint8", "float16"]', '[5, 0, "float32", "float32"]', '[5, 0, "float32", "float16"]', '[5, 0, "float16", "float32"]', '[5, 0, "float16", "float16"]', '[5, 2.0, "uint8", "float32"]', '[5, 2.0, "uint8", "float16"]', '[5, 2.0, "float32", "float32"]', '[5, 2.0, "float32", "float16"]', '[5, 2.0, "float16", "float32"]', '[5, 2.0, "float16", "float16"]', '[5, 4.0, "uint8", "float32"]', '[5, 4.0, "uint8", "float16"]', '[5, 4.0, "float32", "float32"]', '[5, 4.0, "float32", "float16"]', '[5, 4.0, "float16", "float32"]', '[5, 4.0, "float16", "float16"]', '[10, 0, "uint8", "float32"]', '[10, 0, "uint8", "float16"]', '[10, 0, "float32", "float32"]', '[10, 0, "float32", "float16"]', '[10, 0, "float16", "float32"]', '[10, 0, "float16", "float16"]', '[10, 2.0, "uint8", "float32"]', '[10, 2.0, "uint8", "float16"]', '[10, 2.0, "float32", "float32"]', '[10, 2.0, "float32", "float16"]', '[10, 2.0, "float16", "float32"]', '[10, 2.0, "float16", "float16"]', '[10, 4.0, "uint8", "float32"]', '[10, 4.0, "uint8", "float16"]', '[10, 4.0, "float32", "float32"]', '[10, 4.0, "float32", "float16"]', '[10, 4.0, "float16", "float32"]', '[10, 4.0, "float16", "float16"]', '[50, 0, "uint8", "float32"]', '[50, 0, "uint8", "float16"]', '[50, 0, "float32", "float32"]', '[50, 0, "float32", "float16"]', '[50, 0, "float16", "float32"]', '[50, 0, "float16", "float16"]', '[50, 2.0, "uint8", "float32"]', '[50, 2.0, "uint8", "float16"]', '[50, 2.0, "float32", "float32"]', '[50, 2.0, "float32", "float16"]', '[50, 2.0, "float16", "float32"]', '[50, 2.0, "float16", "float16"]', '[50, 4.0, "uint8", "float32"]', '[50, 4.0, "uint8", "float16"]', '[50, 4.0, "float32", "float32"]', '[50, 4.0, "float32", "float16"]', '[50, 4.0, "float16", "float32"]', '[50, 4.0, "float16", "float16"]', '[100, 0, "uint8", "float32"]', '[100, 0, "uint8", "float16"]', '[100, 0, "float32", "float32"]', '[100, 0, "float32", "float16"]', '[100, 0, "float16", "float32"]', '[100, 0, "float16", "float16"]', '[100, 2.0, "uint8", "float32"]', '[100, 2.0, "uint8", "float16"]', '[100, 2.0, "float32", "float32"]', '[100, 2.0, "float32", "float16"]', '[100, 2.0, "float16", "float32"]', '[100, 2.0, "float16", "float16"]', '[100, 4.0, "uint8", "float32"]', '[100, 4.0, "uint8", "float16"]', '[100, 4.0, "float32", "float32"]', '[100, 4.0, "float32", "float16"]', '[100, 4.0, "float16", "float32"]', '[100, 4.0, "float16", "float16"]', '[200, 0, "uint8", "float32"]', '[200, 0, "uint8", "float16"]', '[200, 0, "float32", "float32"]', '[200, 0, "float32", "float16"]', '[200, 0, "float16", "float32"]', '[200, 0, "float16", "float16"]', '[200, 2.0, "uint8", "float32"]', '[200, 2.0, "uint8", "float16"]', '[200, 2.0, "float32", "float32"]', '[200, 2.0, "float32", "float16"]', '[200, 2.0, "float16", "float32"]', '[200, 2.0, "float16", "float16"]', '[200, 4.0, "uint8", "float32"]', '[200, 4.0, "uint8", "float16"]', '[200, 4.0, "float32", "float32"]', '[200, 4.0, "float32", "float16"]', '[200, 4.0, "float16", "float32"]', '[200, 4.0, "float16", "float16"]']
2023-03-15 17:24:25,275 - annb.0a4e95f351 - INFO - ['angular', 400, 5, 64]
2023-03-15 17:24:25,276 - annb.0a4e95f351 - INFO - Trying to instantiate ann_benchmarks.algorithms.raft_ivfpq.RAFTIVFPQ(['angular', 400, 5, 64])
2023-03-15 17:24:31,208 - annb.0a4e95f351 - INFO - got a train set of size (9990000 * 96)
2023-03-15 17:24:31,208 - annb.0a4e95f351 - INFO - got 10000 queries
```

cjnolet commented:
@sourcesync, I just pushed a fix for the error message display. I think after I had finally gotten the docker container to run, I was no longer getting an error so it never tripped on that line.  

I'm curious to know why it failed, though

sourcesync commented:
OK, I will pull the latest changes, rebuild the docker containers, and run
again.



On Wed, Mar 15, 2023 at 3:50 PM Corey J. Nolet ***@***.***>
wrote:

> @sourcesync <https://github.com/sourcesync>, I just pushed a fix for the
> error message display. I think after I had finally gotten the docker
> container to run, I was no longer getting an error so it never tripped on
> that line.
>
> I'm curious to know why it failed, though
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/erikbern/ann-benchmarks/pull/324#issuecomment-1470945108>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CNENMG362WHDH2ON53W4JBR5ANCNFSM6AAAAAAT2WRY6Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sourcesync commented:
Yeah...maybe let's sync on how we are testing in case there are differences
in our environments.  What  commands are you using to 1) build the
container(s) 2) to run the container and on which datasets?

On Wed, Mar 15, 2023 at 5:43 PM Corey J. Nolet ***@***.***>
wrote:

> @sourcesync <https://github.com/sourcesync> I admit I was only testing w/
> the sift-128 dataset. I'll also run the deep-image-96 dataset to make
> sure maybe I didn't set the default memory pool too small.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/erikbern/ann-benchmarks/pull/324#issuecomment-1471085788>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CLBQRO6OTYFBIKYSQ3W4JO47ANCNFSM6AAAAAAT2WRY6Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sourcesync commented:
Cool.  I believe I updated everything correctly. but seeing the following error:
...
(ann-benchmarks) gwilliams@sv7-apu11:~/Projects/ann-benchmarks/cjnolet/ann-benchmarks$ 2023-04-03 22:22:05,209 - annb.b97ca5a7e3ea - INFO - got a train set of size (9000 * 20)
2023-04-03 22:22:05,210 - annb.b97ca5a7e3ea - INFO - got 1000 queries
2023-04-03 22:22:05,213 - annb.b97ca5a7e3ea - INFO - Traceback (most recent call last):
2023-04-03 22:22:05,213 - annb.b97ca5a7e3ea - INFO -   File "/home/app/ann_benchmarks/runner.py", line 133, in run
2023-04-03 22:22:05,213 - annb.b97ca5a7e3ea - INFO -     algo.fit(X_train)
2023-04-03 22:22:05,214 - annb.b97ca5a7e3ea - INFO -   File "/home/app/ann_benchmarks/algorithms/raft_ivfpq.py", line 61, in fit
2023-04-03 22:22:05,214 - annb.b97ca5a7e3ea - INFO -     self._index = ivf_pq.build(index_params, X, handle=self._handle)
2023-04-03 22:22:05,214 - annb.b97ca5a7e3ea - INFO -   File "handle.pyx", line 225, in pylibraft.common.handle.auto_sync_handle.wrapper
2023-04-03 22:22:05,214 - annb.b97ca5a7e3ea - INFO -   File "/opt/conda/envs/rapids/lib/python3.10/site-packages/pylibraft/common/outputs.py", line 83, in wrapper
2023-04-03 22:22:05,214 - annb.b97ca5a7e3ea - INFO -     ret_value = f(*args, **kwargs)
2023-04-03 22:22:05,214 - annb.b97ca5a7e3ea - INFO -   File "ivf_pq.pyx", line 378, in pylibraft.neighbors.ivf_pq.ivf_pq.build
2023-04-03 22:22:05,215 - annb.b97ca5a7e3ea - INFO -   File "ivf_pq.pyx", line 109, in pylibraft.neighbors.ivf_pq.ivf_pq._check_input_array
2023-04-03 22:22:05,215 - annb.b97ca5a7e3ea - INFO - TypeError: 'cai_wrapper' object is not subscriptable
....


My steps are as follows:
* git clone git@github.com:cjnolet/ann-benchmarks.git
* cd ann-benchmarks/
* git checkout add_raft_ivf_pq
* python -m pip install -r requirements.txt

sourcesync commented:
I pulled the latest branch.  I rebuilt containers and I'm getting the following errors:

2023-04-14 16:28:35,407 - annb.b57a66d80d6c - INFO - ['angular', 100, 4, 48]
2023-04-14 16:28:35,408 - annb.b57a66d80d6c - INFO - Trying to instantiate ann_benchmarks.algorithms.raft_ivfpq.RAFTIVFPQ(['angular', 100, 4, 48])
2023-04-14 16:28:38,585 - annb.b57a66d80d6c - INFO - got a train set of size (9000 * 20)
2023-04-14 16:28:38,585 - annb.b57a66d80d6c - INFO - got 1000 queries
2023-04-14 16:28:38,588 - annb.b57a66d80d6c - INFO - Traceback (most recent call last):
2023-04-14 16:28:38,589 - annb.b57a66d80d6c - INFO -   File "/home/app/ann_benchmarks/runner.py", line 139, in run
2023-04-14 16:28:38,589 - annb.b57a66d80d6c - INFO -     algo.fit(X_train)
2023-04-14 16:28:38,589 - annb.b57a66d80d6c - INFO -   File "/home/app/ann_benchmarks/algorithms/raft_ivfpq.py", line 61, in fit
2023-04-14 16:28:38,589 - annb.b57a66d80d6c - INFO -     self._index = ivf_pq.build(index_params, X, handle=self._handle)
2023-04-14 16:28:38,589 - annb.b57a66d80d6c - INFO -   File "handle.pyx", line 225, in pylibraft.common.handle.auto_sync_handle.wrapper
2023-04-14 16:28:38,590 - annb.b57a66d80d6c - INFO -   File "/opt/conda/envs/rapids/lib/python3.10/site-packages/pylibraft/common/outputs.py", line 83, in wrapper
2023-04-14 16:28:38,590 - annb.b57a66d80d6c - INFO -     ret_value = f(*args, **kwargs)
2023-04-14 16:28:38,590 - annb.b57a66d80d6c - INFO -   File "ivf_pq.pyx", line 378, in pylibraft.neighbors.ivf_pq.ivf_pq.build
2023-04-14 16:28:38,590 - annb.b57a66d80d6c - INFO -   File "ivf_pq.pyx", line 109, in pylibraft.neighbors.ivf_pq.ivf_pq._check_input_array
2023-04-14 16:28:38,590 - annb.b57a66d80d6c - INFO - TypeError: 'cai_wrapper' object is not subscriptable

cjnolet commented:
@sourcesync between GTC and PTO, I'm just now getting back to work on this. I realized I had a bad merge last week so I've fixed that and updated the dockerfile to use the (now released) RAPIDS/RAFT 23.04. I'm running tests now on `deep-image-96-angular` with the following command on an RTX A6000: 
```bash
python run.py --dataset deep-image-96-angular --algorithm raft-ivfpq --run-disabled --force --runs 1 --batch --batch-size 1000
```

I'm using this command to build the docker container (please note that you don't have to have nvidia-docker-runtime installed to build the container but it does need to be installed in order to use the container):

```bash
python install.py --algorithm raft
```

So far I haven't seen it fail but I'm going to keep it running and report if it fails. Can you share your build/run commands for your last run? Also, RAPIDS doesn't officially support CUDA 12, though we are hoping to have CUDA 12 support by 23.06. I notice from a previous of run of `nvidia-smi` that you are using CUDA 12. 

Any chance you might be able to use 11.8? In general, the CUDA driver/toolkit should be backwards compatible but there are always weird anomalies that can pop up there so eliminating the potential that those are causing these issues would be helpful. It's hard to tell from your `nvidia-smi` log what GPU you are running- can you share that info as well?

sourcesync commented:
Yeah no problem.   I don't suppose you can get me the exact CUDA installer
link that you want me to use?




On Mon, Apr 17, 2023 at 3:18 PM Corey J. Nolet ***@***.***>
wrote:

> @sourcesync <https://github.com/sourcesync> between GTC and PTO, I'm just
> now getting back to work on this. I realized I had a bad merge last week so
> I've fixed that and updated the dockerfile to use the (now released)
> RAPIDS/RAFT 23.04. I'm running tests now on deep-image-96-angular with
> the following command on an RTX A6000:
>
> python run.py --dataset deep-image-96-angular --algorithm raft-ivfpq --run-disabled --force --runs 1 --batch --batch-size 1000
>
> I'm using this command to build the docker container (please note that you
> don't have to have nvidia-docker-runtime installed to build the container
> but it does need to be installed in order to use the container):
>
> python install.py --algorithm raft
>
> So far I haven't seen it fail but I'm going to keep it running and report
> if it fails. Can you share your build/run commands for your last run? Also,
> RAPIDS doesn't officially support CUDA 12, though we are hoping to have
> CUDA 12 support by 23.06. I notice from a previous of run of nvidia-smi
> that you are using CUDA 12.
>
> Any chance you might be able to use 11.8? In general, the CUDA
> driver/toolkit should be backwards compatible but there are always weird
> anomalies that can pop up there so eliminating the potential that those are
> causing these issues would be helpful. It's hard to tell from your
> nvidia-smi log what GPU you are running- can you share that info as well?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/erikbern/ann-benchmarks/pull/324#issuecomment-1512160023>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CLHU6XEPEJD2RAH67TXBW6S3ANCNFSM6AAAAAAT2WRY6Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cjnolet commented:
This guy right here would be the right one: https://developer.nvidia.com/cuda-11-8-0-download-archive. What card are you using?

sourcesync commented:
> This guy right here would be the right one: https://developer.nvidia.com/cuda-11-8-0-download-archive. What card are you using?

$ nvidia-smi
Mon Apr 17 23:02:22 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.85.12    Driver Version: 525.85.12    CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:3C:00.0 Off |                  N/A |
|  0%   24C    P8     8W / 220W |      5MiB /  8192MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

sourcesync commented:
Thanks.  Ok I should be able to get to it in the next few days...

On Mon, Apr 17, 2023 at 4:00 PM Corey J. Nolet ***@***.***>
wrote:

> This guy right here would be the right one:
> https://developer.nvidia.com/cuda-11-8-0-download-archive. What card are
> you using?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/erikbern/ann-benchmarks/pull/324#issuecomment-1512196498>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADL6CLFFCPPEP25IWGH4Q3XBXDQVANCNFSM6AAAAAAT2WRY6Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cjnolet commented:
Hey @sourcesync just checking in to see if you've had a chance to run this yet? We're in the process of releasing RAPIDS / RAFT 23.06.

sourcesync commented:
> 

Great 👍 I’ll give it a go today…

sourcesync commented:
Wow @cjnolet.  Got a lot further this time.  I'm now getting a memory error.  Do I need to configure something differently?

```
2023-06-06 23:15:27,284 - annb.98409a67adc3 - INFO - /opt/conda/envs/rapids/lib/python3.10/site-packages/rmm/__init__.py:56: FutureWarning: Use of 'rmm.rmm_cupy_allocator' is deprecated and will be removed. 'rmm_cupy_allocator' now lives in the 'rmm.allocators.cupy' sub-module, please update your imports.
2023-06-06 23:15:27,285 - annb.98409a67adc3 - INFO -   warnings.warn(
2023-06-06 23:15:31,125 - annb.98409a67adc3 - INFO - got a train set of size (9990000 * 96)
2023-06-06 23:15:31,125 - annb.98409a67adc3 - INFO - got 10000 queries
2023-06-06 23:15:55,991 - annb.98409a67adc3 - INFO - Built index in 21.052327632904053
2023-06-06 23:15:55,992 - annb.98409a67adc3 - INFO - Index size:  5614504.0
2023-06-06 23:15:55,992 - annb.98409a67adc3 - INFO - Running query argument group 1 of 60...
2023-06-06 23:15:55,992 - annb.98409a67adc3 - INFO - Run 1/1...
2023-06-06 23:15:55,992 - annb.98409a67adc3 - INFO - Running 10 batches with batch size 1000
2023-06-06 23:15:55,997 - annb.98409a67adc3 - INFO - Traceback (most recent call last):
2023-06-06 23:15:55,998 - annb.98409a67adc3 - INFO -   File "/home/app/ann_benchmarks/runner.py", line 155, in run
2023-06-06 23:15:55,998 - annb.98409a67adc3 - INFO -     descriptor, results = run_individual_query(
2023-06-06 23:15:55,998 - annb.98409a67adc3 - INFO -   File "/home/app/ann_benchmarks/runner.py", line 83, in run_individual_query
2023-06-06 23:15:55,998 - annb.98409a67adc3 - INFO -     results = batch_query(X_test, batch_size)
2023-06-06 23:15:55,999 - annb.98409a67adc3 - INFO -   File "/home/app/ann_benchmarks/runner.py", line 71, in batch_query
2023-06-06 23:15:55,999 - annb.98409a67adc3 - INFO -     algo.batch_query(X_batch, count)
2023-06-06 23:15:55,999 - annb.98409a67adc3 - INFO -   File "/home/app/ann_benchmarks/algorithms/raft_ivfpq.py", line 86, in batch_query
2023-06-06 23:15:55,999 - annb.98409a67adc3 - INFO -     D, L = ivf_pq.search(search_params, self._index, X, refine_ratio,
2023-06-06 23:15:55,999 - annb.98409a67adc3 - INFO -   File "handle.pyx", line 225, in pylibraft.common.handle.auto_sync_handle.wrapper
2023-06-06 23:15:56,000 - annb.98409a67adc3 - INFO -   File "/opt/conda/envs/rapids/lib/python3.10/site-packages/pylibraft/common/outputs.py", line 83, in wrapper
2023-06-06 23:15:56,000 - annb.98409a67adc3 - INFO -     ret_value = f(*args, **kwargs)
2023-06-06 23:15:56,000 - annb.98409a67adc3 - INFO -   File "ivf_pq.pyx", line 689, in pylibraft.neighbors.ivf_pq.ivf_pq.search
2023-06-06 23:15:56,000 - annb.98409a67adc3 - INFO -   File "ivf_pq.pyx", line 690, in pylibraft.neighbors.ivf_pq.ivf_pq.search
2023-06-06 23:15:56,000 - annb.98409a67adc3 - INFO - MemoryError: std::bad_alloc: out_of_memory: RMM failure at:/opt/conda/envs/rapids/include/rmm/mr/device/pool_memory_resource.hpp:195: Maximum pool size exceeded
```

cjnolet commented:
My apologies for the delayed response @sourcesync. Yes, this is indeed due to a poor choice of initial pool size [here](https://github.com/erikbern/ann-benchmarks/pull/324/files#diff-734c08589fa1cd53bc4c1f6844921f329415eda452c0633cac457f39375ec338R33). I think we can remove the max pool size setting altogether.

sourcesync commented:
No worries @cjnolet.  Let me know if you want me to test again.

cjnolet commented:
@sourcesync I adjusted the pooled allocator to start w/ 4gb and grow as needed (rather than capping at 8gb). If you are able to give it another try, it should hopefully work now. 

In the meantime, I see there were some significant changes to the codebase configuration so I'll start looking at fixing those merge conflicts. Hopefully we can get this in soon before it diverges too much further.

cjnolet commented:
@sourcesync any change you've been able to give this another go?

sourcesync commented:
Sorry @cjnolet I lost track and didn't realize the ball was in my court.  Yeah, i'll need some time.  I'll probably get to it tomorrow definitely by this weekend.

sourcesync commented:
Hey @cjnolet.  It ran for a while and then abruptly terminated with what is likely OOM...

```
2023-07-24 22:18:15,126 - annb.187718d328f4 - INFO - Running 10 batches with batch size 1000                                                                                                       2023-07-24 22:18:18,556 - annb.187718d328f4 - INFO - Running query argument group 56 of 60...                                                                                                      2023-07-24 22:18:18,556 - annb.187718d328f4 - INFO - Run 1/1...                                                                                                                                    2023-07-24 22:18:18,556 - annb.187718d328f4 - INFO - Running 10 batches with batch size 1000                                                                                                       2023-07-24 22:18:21,924 - annb.187718d328f4 - INFO - Running query argument group 57 of 60...                                                                                                      2023-07-24 22:18:21,924 - annb.187718d328f4 - INFO - Run 1/1...                                                                                                                                    2023-07-24 22:18:21,924 - annb.187718d328f4 - INFO - Running 10 batches with batch size 1000                                                                                                       2023-07-24 22:18:25,283 - annb.187718d328f4 - INFO - Running query argument group 58 of 60...                                                                                                      2023-07-24 22:18:25,284 - annb.187718d328f4 - INFO - Run 1/1...                                                                                                                                    2023-07-24 22:18:25,284 - annb.187718d328f4 - INFO - Running 10 batches with batch size 1000                                                                                                       2023-07-24 22:18:28,592 - annb.187718d328f4 - INFO - Running query argument group 59 of 60...                                                                                                      2023-07-24 22:18:28,592 - annb.187718d328f4 - INFO - Run 1/1...                                                                                                                                    2023-07-24 22:18:28,592 - annb.187718d328f4 - INFO - Running 10 batches with batch size 1000                                                                                                       2023-07-24 22:18:32,056 - annb.187718d328f4 - INFO - Running query argument group 60 of 60...                                                                                                      2023-07-24 22:18:32,057 - annb.187718d328f4 - INFO - Run 1/1...                                                                                                                                    2023-07-24 22:18:32,057 - annb.187718d328f4 - INFO - Running 10 batches with batch size 1000                                                                                                       
```
I ran this command:
```
python run.py --dataset deep-image-96-angular --algorithm raft-ivfpq --run-disabled --force --runs 1 --batch --batch-size 1000
```

cjnolet commented:
Thanks for running this again, @sourcesync. I'm not sure it failed because I'd have expected to see an exception if it did. 

Can you please share the command you used to generate the plots?

sourcesync commented:
> Thanks for running this again, @sourcesync. I'm not sure it failed because I'd have expected to see an exception if it did.
> 
> I believe the `--batch` (and `--batch-size`) flags also need to be used in the call to `plot.py`.

Yeah let try that ( you are right, no error )...

sourcesync commented:
@cjnolet alright, sorry for the false alarm on the run error.  I reran plotting via...

```
python plot.py --dataset deep-image-96-angular  --batch --batch-size 1000
```

And got the attached plot!  

![deep-image-96-angular-batch](https://github.com/erikbern/ann-benchmarks/assets/442121/44d52768-d8ce-4198-bd39-45a578c4e929)

cjnolet commented:
Ah ha. This is great! Thanks @sourcesync! Okay, I'll spend some time merging the newer changes into this PR. We will soon be releasing a python API for our new graph-base method as well (called CAGRA), which is optimized for the smaller query batches (1-100) on GPU.

sourcesync commented:
> Ah ha. This is great! Thanks @sourcesync! Okay, I'll spend some time merging the newer changes into this PR. We will soon be releasing a python API for our new graph-base method as well (called CAGRA), which is optimized for the smaller query batches (1-100) on GPU.

Wow, that's super cool @cjnolet   Will you be publishing any papers or blogs on those topics?  Anyway, let me know if you need another test, but looking good right now.

cjnolet commented:
@sourcesync yes we are currently finishing up a paper on CAGRA and I'll definitely share that when it's ready. 

We also have a blog being published that includes CAGRA. I'll definitely share that as well!

sourcesync commented:
Great @cjnolet feel free to send over any drafts my way.  Happy to help, looks like you guys have something super interesting ( I also help organize NeurIPS ANN competition. )

sourcesync commented:
Hey @cjnolet.  Just following up on our thread.  I'm curious - will you guys be at NEURIPs this year?

cjnolet commented:
Sorry guys, I've been intending to finish up these changes. We have just been very busy on the raft side lately. What's left of this PR is just reconcile the changes with changes that have been made to Ann-benchmarks since I began the Pr.

@erikbern would you prefer I open a new PR when ready instead of keeping this one open?

cjnolet commented:
@sourcesync my apologies for the delayed response. We've been working hard on the Nvidia side to make our GPU-accelerated vector search capabilities more fully featured (working on things like CRUD, interop between GPU<->CPU, multi-gpu, out of core, etc...). 

Here's a pre-print of the [CAGRA paper](https://arxiv.org/abs/2308.15136). I'm not going to be at Neurips this year but I'd love to sync up sometime and chat about this stuff, as well as GPU-accelerated vector search in general. 

I'd still love to get the RAFT indexes integrated into ANN benchmarks (and enable the larger batch sizes like this PR was doing). Are you all still interested in such a change? cc @erikbern

sourcesync commented:
> @sourcesync my apologies for the delayed response. We've been working hard on the Nvidia side to make our GPU-accelerated vector search capabilities more fully featured (working on things like CRUD, interop between GPU<->CPU, multi-gpu, out of core, etc...).
> 
> Here's a pre-print of the [CAGRA paper](https://arxiv.org/abs/2308.15136). I'm not going to be at Neurips this year but I'd love to sync up sometime and chat about this stuff, as well as GPU-accelerated vector search in general.
> 
> I'd still love to get the RAFT indexes integrated into ANN benchmarks (and enable the larger batch sizes like this PR was doing). Are you all still interested in such a change? cc @erikbern

No worries @cjnolet.  Actually, a had different question for you.  A few of us are hosting a NEURIPs workshop this year and we are inviting speakers - your work with RAFT and CARGA would make for a great talk.  By the way, I know Akira and Yong from their participation a couple years ago in the hardware track for our first NEURIPs ANN challenge.  I think Frank Lui and Edo Liberty have also reached out to you on the same topic.

cjnolet commented:
> A few of us are hosting a NEURIPs workshop this year and we are inviting speakers - your work with RAFT and CARGA would make for a great talk. 

This sounds great. I'd love to learn more about it. 

> I think Frank Lui and Edo Liberty have also reached out to you on the same topic.

Oh, I had no clue. Do you want to take this offline? Have you been reaching out to me specifically or other members of the Nvidia / RAFT team?
